File: src/main/java/chapter02/ReverseList.java
package chapter02;

import utils.ListNode;

/**
 * <h2>反转链表</h2>
 * <h3>1.反转链表</h3>
 * <h3>2.反转链表 II</h3>
 * <h3>3.两两交换链表中的结点</h3>
 * <h3>4.K个一组反转</h3>
 * <p>万能解法: 集合、数组、栈都是可以完成的</p>
 */
public class ReverseList {

    /**
     * <h3>思路: 反转链表</h3>
     */
    private static ListNode reverse(ListNode head){
        ListNode previous = null, current = head, next = null;
        while (current != null){
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        return previous;
    }

    /**
     * <h3>思路: 反转链表 II</h3>
     */
    public static ListNode reverseBetween(ListNode head, int left, int right) {
        int index = 1;
        // 注: 这个变量的目的是为了记录左边界的前驱结点
        ListNode dummy = new ListNode(0, head);
        ListNode previous = null, current = head, next = null;
        // 1. 找到左边界之后就可以开始反转链表了
        while (index <= left){
            // 2. 记录左边界的前驱结点, 这么处理的原因是可能恰好反转的部分是从头开始的
            dummy = index == left - 1 ? current: dummy;
            // 3. 反转链表的前驱结点会移动到左边界
            previous = current;
            current = current.next;
            index++;
        }
        // 4. 从左边界开始反转链表
        while (index <= right){
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
            index++;
        }
        // 5. 通过左边界前驱结点修改左边界的后继结点
        dummy.next.next = current;
        // 6. 修改前驱结点的后继结点
        dummy.next = previous;
        // 7. 如果左边界前驱结点仍然是哑元, 那么证明反转就是从头结点开始的
        return dummy.value != 0 ? head: dummy.next;
    }

    /**
     * <h3>思路: K个一组反转链表</h3>
     */
    public static ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0, head);
        ListNode tail = dummy;
        ListNode previous = dummy;
        ListNode current = head;
        ListNode next = null;
        while(current != null){
            // 找到需要反转的区域, 也兼具查看剩下的链表是否需要反转的功能
            for(int index = 0;index < k;index++){
                tail = tail.next;
                // 如果尾指针为空, 证明剩余部分不够
                if(tail == null) return dummy.next;
            }
            // 记录下一部分链表的头结点
            next = tail.next;
            // 反转链表
            ListNode[] nodes = reverse(current, tail);
            // 接回原链表
            current = nodes[0];
            tail = nodes[1];
            previous.next = current;
            tail.next = next;
            // 更新尾指针
            previous = tail;
            // 向后移动
            current = next;
        }
        return dummy.next;
    }

    public static ListNode[] reverse(ListNode head, ListNode tail){
        ListNode previous = null;
        ListNode current = head;
        ListNode next = null;
        ListNode end = tail.next;
        while(current != end){
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        return new ListNode[]{tail, head};
    }
}

File: leetcode-go/chapter02/reverse_list.go
package main

// 反转链表: 迭代
func reverseListV1(head *ListNode) *ListNode {
	var previous *ListNode = nil
	var current *ListNode = head
	for current != nil {
		next := current.Next
		current.Next = previous
		previous = current
		current = next
	}
	return previous
}

// 反转链表: 递归
func reverseListV2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	tail := reverseListV2(head.Next)
	head.Next.Next = head
	head.Next = nil
	return tail
}

// 反转链表 II
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	dummy := &ListNode{0, head}
	leftBound := dummy
	// 1. 找到左右边界的前驱和后驱节点
	for index := 0; index < left - 1; index++ {
		leftBound = leftBound.Next
	}
	rightBound := leftBound
	for index := left - 1; index < right; index++ {
		rightBound = rightBound.Next
	}
	// 2. 断开左右边界的前驱和后驱节点
	start, end := leftBound.Next, rightBound.Next
	leftBound.Next, rightBound.Next = nil, nil
	// 3. 开始反转
	var previous *ListNode = nil
	var current *ListNode = start
	for current != nil {
		next := current.Next
		current.Next = previous
		previous = current
		current = next
	}
	leftBound.Next = previous
	start.Next = end
	return dummy.Next
}

// 反转链表: 两两交换
// 方法一：模拟
// 方法二：套 K 个一组的解法
func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	current := dummy
	for current.Next != nil && current.Next.Next != nil {
		node1, node2 := current.Next, current.Next.Next
		current.Next = node2
		node1.Next = node2.Next
		node2.Next = node1
		current = node2.Next
	}
	return dummy.Next
}

// 反转链表: K 个一组反转
func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	previous, current := dummy, head
	for current != nil {
		leftBound, rightBound := previous, previous
		// 找到链接的前驱和后继结点
		for index := 0; index < k; index++ {
			rightBound = rightBound.Next
			if rightBound == nil {
				return dummy.Next
			}
		}
		// 断开前驱后继结点
		start, end := leftBound.Next, rightBound.Next
		leftBound.Next, rightBound.Next = nil, nil
		// 反转链表
		for current != nil {
			next := current.Next
			current.Next = previous
			previous = current
			current = next
		}
		leftBound.Next = previous
		start.Next = end
		previous, current = start, end
	}
	return dummy.Next
}
